Index: phpbb_seo/phpbb_seo_class.php
===================================================================
--- phpbb_seo/phpbb_seo_class.php	(0.4.4)
+++ phpbb_seo/phpbb_seo_class.php	(0.4.6)
@@ -13,7 +13,7 @@
 * @package Advanced phpBB3 SEO mod Rewrite
 */
 class phpbb_seo {
-	var	$version = '0.4.4';
+	var	$version = '0.4.6';
 	var	$modrtype = 0;
 	var	$seo_paths = array();
 	var	$seo_url = array();
@@ -33,8 +33,10 @@
 	var	$seo_opt = array();
 	var	$rewrite_functions = array();
 	var	$seo_cache = array();
-	var	$encoding = 'utf-8';
 	var	$cache_config = array();
+	var	$seo_stop_files = array();
+	var	$seo_stop_vars = array();
+	var	$seo_stop_dirs = array();
 	/**
 	* constuctor
 	*/
@@ -57,7 +59,9 @@
 		// Stop files
 		$this->seo_stop_files = array('posting' => 1, 'faq' => 1, 'ucp' => 1, 'swatch' => 1, 'mcp' => 1);
 		// Stop vars
-		$this->seo_stop_vars = array('view=', 'mark=', 'watch=');
+		$this->seo_stop_vars = array('view=', 'mark=', 'watch=', 'hash=');
+		// Stop dirs
+		$this->seo_stop_dirs = array($phpbb_root_path . 'adm/' => false,);
 		// reset GET var array
 		$this->get_vars = array();
 		// Delimiters : used as separators in the .htaccess RegEx
@@ -161,9 +165,10 @@
 			$this->seo_opt['url_pattern'][] = '`(^|-)[a-z0-9]{1,2}(?=-|$)`i';
 		}
 		$this->seo_opt['url_pattern'][] ='`[-]+`'; // Do not remove : multi hyphen reduction
-		// Rewrite functions array : array('file_name' => 'function_name');
-		// Allow to add options without slowing down the URL rewrite process
-		$this->rewrite_functions = array( 'viewtopic' => $this->seo_opt['virtual_folder'] ? 'viewtopic_uadv' : 'viewtopic_adv',
+		// Rewrite functions array : array( 'path' => array('file_name' => 'function_name'));
+		// Warning, this way of doing things is path aware, this implies path to be properly sent to append_sid()
+		// Allow to add options without slowing down the URL rewriting process
+		$this->rewrite_functions[$phpbb_root_path] = array( 'viewtopic' => $this->seo_opt['virtual_folder'] ? 'viewtopic_uadv' : 'viewtopic_adv',
 			'viewforum' => 'viewforum_adv',
 			'index' => 'index',
 			'memberlist' => $this->seo_opt['profile_inj'] ? 'memberlist_adv' : 'memberlist_smpl',
@@ -199,16 +204,15 @@
 		$this->seo_ext['index'] = empty($this->seo_static['index']) ? '' : ( empty($this->seo_ext['index']) ? '.html' : $this->seo_ext['index']);
 		// File setting
 		$this->seo_req_uri();
-		$this->seo_opt['seo_base_href'] = $this->seo_opt['req_file'] = '';
+		$this->seo_opt['seo_base_href'] = $this->seo_opt['req_file'] = $this->seo_opt['req_self'] = '';
 		if ($script_name = (!empty($_SERVER['PHP_SELF'])) ? $_SERVER['PHP_SELF'] : getenv('PHP_SELF')) {
 			// From sessions.php
 			// Replace backslashes and doubled slashes (could happen on some proxy setups)
-			$script_name = str_replace(array('\\', '//'), '/', $script_name);
-			// basenamed page name (for example: index.php)
-			$script_name = urlencode(htmlspecialchars(basename($script_name)));
-			$this->seo_opt['req_file'] = str_replace('.' . $phpEx, '', $script_name);
+			$this->seo_opt['req_self'] = str_replace(array('\\', '//'), '/', $script_name);
+			// basenamed page name (for example: index)
+			$this->seo_opt['req_file'] = urlencode(htmlspecialchars(str_replace('.' . $phpEx, '', basename($this->seo_opt['req_self']))));
 		}
-		if ( !defined('ADMIN_START') && isset($this->seo_opt['file_hbase'][$this->seo_opt['req_file']])) {	
+		if ( $this->seo_opt['url_rewrite'] && !defined('ADMIN_START') && isset($this->seo_opt['file_hbase'][$this->seo_opt['req_file']])) {	
 			$this->seo_opt['seo_base_href'] = '<base href="' . $this->seo_opt['file_hbase'][$this->seo_opt['req_file']] . '"/>';
 		}
 		return;
@@ -219,7 +223,7 @@
 	*/
 	function format_url( $url, $type = 'topic' ) {
 		$url = preg_replace('`\[.*\]`U','',$url);
-		$url = htmlentities($url, ENT_COMPAT, $this->encoding);
+		$url = htmlentities($url, ENT_COMPAT, 'utf-8');
 		$url = preg_replace( '`&([a-z]+)(acute|uml|circ|grave|ring|cedil|slash|tilde|caron|lig);`i', "\\1", $url );
 		$url = preg_replace( $this->seo_opt['url_pattern'] , '-', $url);
 		$url = strtolower(trim($url, '-'));
@@ -232,6 +236,22 @@
 		return (!empty($this->cache_config[$type][$id])) ? $this->cache_config[$type][$id] : $this->format_url( $url, $type ) . $this->seo_delim[$type] . $id;
 	}
 	/**
+	* Prepare url first part
+	*/
+	function prepare_url( $type, $title, $id ) {
+		if ( empty($this->seo_url[$type][$id]) ) {
+			$this->seo_url[$type][$id] = $this->format_url($title, $this->seo_static[$type]) . $this->seo_delim[$type] . $id;
+		}
+	}
+	/**
+	* Set title
+	*/
+	function set_title( $type, $title, $id ) {
+		if ( empty($this->seo_url[$type][$id]) ) {
+			$this->seo_url[$type][$id] = $this->format_url($title, $this->seo_static[$type]);
+		}
+	}
+	/**
 	* Prepare profile url
 	*/
 	function set_user_url( $username, $user_id = 0 ) {
@@ -265,26 +285,26 @@
 	* Mimics append_sid with some shortcuts related to how url are rewritten
 	*/
 	function url_rewrite($url, $params = false, $is_amp = true, $session_id = false) {
-		global $phpEx, $user, $_SID, $_EXTRA_URL;
+		global $phpEx, $user, $_SID, $_EXTRA_URL, $phpbb_root_path;
 		$qs = $anchor = '';
 		$amp_delim = ($is_amp) ? '&amp;' : '&';
-		@list($this->path, $qs) = explode('?', $url, 2);
 		if (strpos($url, '#') !== false) {
 			list($url, $anchor) = explode('#', $url, 2);
 			$anchor = '#' . $anchor;
 		}
+		@list($this->path, $qs) = explode('?', $url, 2);
 		if (is_array($params)) {
 			if (!empty($params['#'])) {
 				$anchor = '#' . $params['#'];
 				unset($params['#']);
 			}
-			$qs = $this->query_string($params, $amp_delim, '');
+			$qs .= ($qs ? $amp_delim : '') . $this->query_string($params, $amp_delim, '');
 		} elseif ($params) {
 			if (strpos($params, '#') !== false) {
 				list($params, $anchor) = explode('#', $params, 2);
 				$anchor = '#' . $anchor;
 			}
-			$qs = $params;
+			$qs .= ($qs ? $amp_delim : '') . $params;
 		}
 		// Appending custom url parameter?
 		if (!empty($_EXTRA_URL)) {
@@ -297,13 +317,19 @@
 			$qs .= ($qs ? $amp_delim : '') . "sid=$_SID";
 		}
 		// Build vanilla URL
-		$this->file = basename($this->path);
-		$this->path = str_replace($this->file, '', $this->path);
-		$url = $this->url_in = $this->path . $this->file . ($qs ? '?' . $qs : '');
+		if (preg_match("`\.[a-z0-9]+$`i", $this->path) ) {
+			$this->file = basename($this->path);
+			$this->path = ltrim(str_replace($this->file, '', $this->path), '/');
+		} else {
+			$this->file = '';
+			$this->path = ltrim($this->path, '/');
+		}
+		$this->url_in = $this->file . ($qs ? '?' . $qs : '');
+		$url = $this->path . $this->url_in . $anchor;
 		if (isset($this->seo_cache[$url])) {
 			return $this->seo_cache[$url];
 		}
-		if ( !$this->seo_opt['url_rewrite'] || strpos($this->path, 'adm/') !== false || defined('ADMIN_START') ) {
+		if ( !$this->seo_opt['url_rewrite'] || defined('ADMIN_START') || isset($this->seo_stop_dirs[$this->path]) ) {
 			return ($this->seo_cache[$url] = $url);
 		}	
 		$this->filename = trim(str_replace(".$phpEx", '', $this->file));
@@ -320,8 +346,8 @@
 			}
 		}
 		$this->url = $this->file;
-		if ( !empty($this->rewrite_functions[$this->filename]) ) {
-			$this->{$this->rewrite_functions[$this->filename]}();
+		if ( !empty($this->rewrite_functions[$this->path][$this->filename]) ) {
+			$this->{$this->rewrite_functions[$this->path][$this->filename]}();
 			return ($this->seo_cache[$url] = $this->path . $this->url . $this->query_string($this->get_vars, $amp_delim, '?') . $anchor);
 		} else {
 			return ($this->seo_cache[$url] = $url);
@@ -333,6 +359,7 @@
 	* @access private
 	*/
 	function viewtopic_uadv() {
+		global $phpbb_root_path;
 		$this->filter_url($this->seo_stop_vars);
 		$this->path = $this->seo_path['phpbb_urlR'];
 		if ( !empty($this->get_vars['p']) ) {
@@ -340,10 +367,10 @@
 			unset($this->get_vars['p'], $this->get_vars['f'], $this->get_vars['t'], $this->get_vars['start']);
 			return;
 		}
-		if ( !empty($this->get_vars['t']) && !empty($this->get_vars['f']) && isset($this->seo_url['topic'][$this->get_vars['t']]) && isset($this->seo_url['forum'][$this->get_vars['f']])) {
+		if ( !@empty($this->seo_url['topic'][$this->get_vars['t']]) && !@empty($this->seo_url['forum'][$this->get_vars['f']])) {
 			// Filter default params
 			$this->filter_get_var($this->phpbb_filter['topic']);
-			$this->{$this->rewrite_functions['topic_pagination']}($this->seo_ext['topic']);
+			$this->{$this->rewrite_functions[$phpbb_root_path]['topic_pagination']}($this->seo_ext['topic']);
 			$this->url = $this->seo_url['topic'][$this->get_vars['t']] . $this->seo_delim['topic'] . $this->get_vars['t'] . $this->start;
 			if (@$this->seo_opt['topic_type'][$this->get_vars['t']] == POST_GLOBAL) {
 				$this->url = $this->seo_static['global_announce'] . $this->seo_ext['global_announce'] . $this->url;
@@ -362,6 +389,7 @@
 	* @access private
 	*/
 	function viewtopic_adv() {
+		global $phpbb_root_path;
 		$this->filter_url($this->seo_stop_vars);
 		$this->path = $this->seo_path['phpbb_urlR'];
 		if ( !empty($this->get_vars['p']) ) {
@@ -369,10 +397,10 @@
 			unset($this->get_vars['p'], $this->get_vars['f'], $this->get_vars['t'], $this->get_vars['start']);
 			return;
 		}
-		if ( !empty($this->get_vars['t']) && !empty($this->seo_url['topic'][$this->get_vars['t']]) ) {
+		if ( !@empty($this->seo_url['topic'][$this->get_vars['t']]) ) {
 			// Filter default params
 			$this->filter_get_var($this->phpbb_filter['topic']);
-			$this->{$this->rewrite_functions['topic_pagination']}($this->seo_ext['topic']);
+			$this->{$this->rewrite_functions[$phpbb_root_path]['topic_pagination']}($this->seo_ext['topic']);
 			$this->url = $this->seo_url['topic'][$this->get_vars['t']] . $this->seo_delim['topic'] . $this->get_vars['t'] . $this->start;
 			unset($this->get_vars['t'], $this->get_vars['f'],$this->get_vars['p']);
 			return;
@@ -385,12 +413,13 @@
 	* @access private
 	*/
 	function viewforum_adv() {
+		global $phpbb_root_path;
 		$this->path = $this->seo_path['phpbb_urlR'];
 		$this->filter_url($this->seo_stop_vars);
-		if ( !empty($this->get_vars['f']) && isset($this->seo_url['forum'][$this->get_vars['f']]) ) {
+		if ( !@empty($this->seo_url['forum'][$this->get_vars['f']]) ) {
 			// Filter default params
 			$this->filter_get_var($this->phpbb_filter['forum']);
-			$this->{$this->rewrite_functions['forum_pagination']}($this->seo_ext['forum']);
+			$this->{$this->rewrite_functions[$phpbb_root_path]['forum_pagination']}($this->seo_ext['forum']);
 			$this->url = $this->seo_url['forum'][$this->get_vars['f']] . $this->start;
 			unset($this->get_vars['f']);
 			return;
@@ -404,13 +433,14 @@
 	* @access private
 	*/
 	function memberlist_adv() {
+		global $phpbb_root_path;
 		$this->path = $this->seo_path['phpbb_urlR'];
-		if ( !empty($this->get_vars['u']) && @$this->get_vars['mode'] === 'viewprofile' && isset($this->seo_url['user'][$this->get_vars['u']]) ) {
+		if ( @$this->get_vars['mode'] === 'viewprofile' && !@empty($this->seo_url['user'][$this->get_vars['u']]) ) {
 			$this->url = $this->seo_url['user'][$this->get_vars['u']] . $this->seo_ext['user'];
 			unset($this->get_vars['mode'], $this->get_vars['u']);
 			return;
-		} elseif ( !empty($this->get_vars['g']) && @$this->get_vars['mode'] === 'group' && isset($this->seo_url['group'][$this->get_vars['g']]) ) {
-			$this->{$this->rewrite_functions['group_pagination']}($this->seo_ext['group']);
+		} elseif ( @$this->get_vars['mode'] === 'group' && !@empty($this->seo_url['group'][$this->get_vars['g']]) ) {
+			$this->{$this->rewrite_functions[$phpbb_root_path]['group_pagination']}($this->seo_ext['group']);
 			$this->url =  $this->seo_url['group'][$this->get_vars['g']] . $this->seo_delim['group'] . $this->get_vars['g'] . $this->start;
 			unset($this->get_vars['mode'], $this->get_vars['g']);
 			return;
@@ -428,13 +458,14 @@
 	* @access private
 	*/
 	function memberlist_smpl() {
+		global $phpbb_root_path;
 		$this->path = $this->seo_path['phpbb_urlR'];
 		if ( !empty($this->get_vars['u']) && @$this->get_vars['mode'] === 'viewprofile' ) {
 			$this->url =  $this->seo_static['user'] . $this->get_vars['u'] . $this->seo_ext['user'];
 			unset($this->get_vars['mode'], $this->get_vars['u']);
 			return;
 		} elseif ( !empty($this->get_vars['g']) && @$this->get_vars['mode'] === 'group' ) {
-			$this->{$this->rewrite_functions['group_pagination']}($this->seo_ext['group']);
+			$this->{$this->rewrite_functions[$phpbb_root_path]['group_pagination']}($this->seo_ext['group']);
 			$this->url = $this->seo_static['group'] . $this->get_vars['g'] . $this->start;
 			unset($this->get_vars['mode'], $this->get_vars['g']);
 			return;
@@ -451,12 +482,13 @@
 	* @access private
 	*/
 	function search_smpl() {
+		global $phpbb_root_path;
 		$this->path = $this->seo_path['phpbb_urlR'];
 		$user_id = !empty($this->get_vars['author_id']) ? $this->get_vars['author_id'] : ( isset($this->seo_url['username'][rawurldecode(@$this->get_vars['author'])]) ? $this->seo_url['username'][rawurldecode($this->get_vars['author'])] : 0);
 		if ( $user_id ) {
 			// Filter default params
 			$this->filter_get_var($this->phpbb_filter['search']);
-			$this->{$this->rewrite_functions['user_pagination']}($this->seo_ext['user']);
+			$this->{$this->rewrite_functions[$phpbb_root_path]['user_pagination']}($this->seo_ext['user']);
 			$sr = (@$this->get_vars['sr'] == 'topics' ) ? 'topics' : 'posts';
 			$this->url = $this->seo_static['user'] . $user_id . $this->seo_delim['sr'] . $sr . $this->start;
 			unset($this->get_vars['author_id'], $this->get_vars['author'], $this->get_vars['sr']);
@@ -465,7 +497,7 @@
 			switch ($this->get_vars['search_id']) {
 				case 'active_topics':
 					$this->filter_get_var($this->phpbb_filter['search']);
-					$this->{$this->rewrite_functions['atopic_pagination']}($this->seo_ext['atopic']);
+					$this->{$this->rewrite_functions[$phpbb_root_path]['atopic_pagination']}($this->seo_ext['atopic']);
 					$this->url = $this->seo_static['atopic'] . $this->start;
 					unset($this->get_vars['search_id'], $this->get_vars['sr']);
 					if (@$this->get_vars['st'] == 7) {
@@ -474,7 +506,7 @@
 					return;
 				case 'unanswered':
 					$this->filter_get_var($this->phpbb_filter['search']);
-					$this->{$this->rewrite_functions['utopic_pagination']}($this->seo_ext['utopic']);
+					$this->{$this->rewrite_functions[$phpbb_root_path]['utopic_pagination']}($this->seo_ext['utopic']);
 					$this->url = $this->seo_static['utopic'] . $this->start;
 					unset($this->get_vars['search_id']);
 					if (@$this->get_vars['sr'] == 'topics') {
@@ -488,7 +520,7 @@
 					return;
 				case 'newposts':
 					$this->filter_get_var($this->phpbb_filter['search']);
-					$this->{$this->rewrite_functions['npost_pagination']}($this->seo_ext['npost']);
+					$this->{$this->rewrite_functions[$phpbb_root_path]['npost_pagination']}($this->seo_ext['npost']);
 					$this->url = $this->seo_static['npost'] . $this->start;
 					unset($this->get_vars['search_id']);
 					if (@$this->get_vars['sr'] == 'topics') {
@@ -505,12 +537,13 @@
 	* @access private
 	*/
 	function search_adv() {
+		global $phpbb_root_path;
 		$this->path = $this->seo_path['phpbb_urlR'];
 		$user_id = !empty($this->get_vars['author_id']) ? $this->get_vars['author_id'] : ( isset($this->seo_url['username'][rawurldecode(@$this->get_vars['author'])]) ? $this->seo_url['username'][rawurldecode($this->get_vars['author'])] : 0);
 		if ( $user_id && isset($this->seo_url['user'][$user_id]) ) {
 			// Filter default params
 			$this->filter_get_var($this->phpbb_filter['search']);
-			$this->{$this->rewrite_functions['user_pagination']}($this->seo_ext['user']);
+			$this->{$this->rewrite_functions[$phpbb_root_path]['user_pagination']}($this->seo_ext['user']);
 			$sr = (@$this->get_vars['sr'] == 'topics' ) ? 'topics' : 'posts';
 			$this->url = $this->seo_url['user'][$user_id] . $this->seo_delim['sr'] . $sr . $this->start;
 			unset($this->get_vars['author_id'], $this->get_vars['author'], $this->get_vars['sr']);
@@ -527,7 +560,7 @@
 			switch ($this->get_vars['search_id']) {
 				case 'active_topics':
 					$this->filter_get_var($this->phpbb_filter['search']);
-					$this->{$this->rewrite_functions['atopic_pagination']}($this->seo_ext['atopic']);
+					$this->{$this->rewrite_functions[$phpbb_root_path]['atopic_pagination']}($this->seo_ext['atopic']);
 					$this->url = $this->seo_static['atopic'] . $this->start;
 					unset($this->get_vars['search_id'], $this->get_vars['sr']);
 					if (@$this->get_vars['st'] == 7) {
@@ -536,7 +569,7 @@
 					return;
 				case 'unanswered':
 					$this->filter_get_var($this->phpbb_filter['search']);
-					$this->{$this->rewrite_functions['utopic_pagination']}($this->seo_ext['utopic']);
+					$this->{$this->rewrite_functions[$phpbb_root_path]['utopic_pagination']}($this->seo_ext['utopic']);
 					$this->url = $this->seo_static['utopic'] . $this->start;
 					unset($this->get_vars['search_id']);
 					if (@$this->get_vars['sr'] == 'topics') {
@@ -551,7 +584,7 @@
 					return;
 				case 'newposts':
 					$this->filter_get_var($this->phpbb_filter['search']);
-					$this->{$this->rewrite_functions['npost_pagination']}($this->seo_ext['npost']);
+					$this->{$this->rewrite_functions[$phpbb_root_path]['npost_pagination']}($this->seo_ext['npost']);
 					$this->url = $this->seo_static['npost'] . $this->start;
 					unset($this->get_vars['search_id']);
 					if (@$this->get_vars['sr'] == 'topics') {
