Index: setup_phpbb_seo.php
===================================================================
--- setup_phpbb_seo.php	(0.6.4)
+++ setup_phpbb_seo.php	(0.6.6)
@@ -2,7 +2,7 @@
 /**
 *
 * @package Ultimate SEO URL phpBB SEO
-* @version $Id: setup_phpbb_seo.php 222 2010-02-27 13:08:48Z dcz $
+* @version $Id: setup_phpbb_seo.php 262 2010-04-20 11:06:58Z dcz $
 * @copyright (c) 2006 - 2010 www.phpbb-seo.com
 * @license http://www.opensource.org/licenses/rpl1.5.txt Reciprocal Public License 1.5
 *

